Apply saved tensor hooks

Varal7 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#60975**

Summary: Fixes #58512
Uses the hooks introduced in #60663: upon registering, the pack hook is called and the returned python object is stored. From then on, whenever we need to unpack it, we will use that python object in combination with the unpack hook.
The packing can be done with gradient tracking disabled as we will add back the correct grad_fn during unpacking.
Inplace operations done by the pack_hook on the original tensor (in the case `leaf || !output`) will be caught if the Saved tensor is used by another op.
Inplace operations done by unpack_hook will unfortunately not be caught. We will add a warning in the docs that follows this PR.


Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29466227](https://our.internmc.facebook.com/intern/diff/D29466227)

Varal7 committed:
Apply saved tensor hooks

Summary: Fixes #58512

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29466227).

Varal7 committed:
Update on "Apply saved tensor hooks"


Summary: Fixes #58512

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29466227](https://our.internmc.facebook.com/intern/diff/D29466227)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29466227).

Varal7 committed:
Update on "Apply saved tensor hooks"


Summary: Fixes #58512

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29466227](https://our.internmc.facebook.com/intern/diff/D29466227)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29466227).

Varal7 committed:
Update on "Apply saved tensor hooks"


Summary: Fixes #58512

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29466227](https://our.internmc.facebook.com/intern/diff/D29466227)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29466227).

Varal7 committed:
Update on "Apply saved tensor hooks"


Summary: Fixes #58512

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29466227](https://our.internmc.facebook.com/intern/diff/D29466227)

[ghstack-poisoned]

Varal7 committed:
Update on "Apply saved tensor hooks"


Summary: Fixes #58512

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29466227](https://our.internmc.facebook.com/intern/diff/D29466227)

[ghstack-poisoned]

Varal7 committed:
Update on "Apply saved tensor hooks"


Summary: Fixes #58512

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29466227](https://our.internmc.facebook.com/intern/diff/D29466227)

[ghstack-poisoned]

Varal7 committed:
Update on "Apply saved tensor hooks"


Summary: Fixes #58512

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29466227](https://our.internmc.facebook.com/intern/diff/D29466227)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29466227).

Varal7 committed:
Update on "Apply saved tensor hooks"


Summary: Fixes #58512

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29466227](https://our.internmc.facebook.com/intern/diff/D29466227)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29466227).

Varal7 committed:
Update on "Apply saved tensor hooks"


Summary: Fixes #58512

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29466227](https://our.internmc.facebook.com/intern/diff/D29466227)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29466227).

Varal7 committed:
Update on "Apply saved tensor hooks"


Summary: Fixes #58512

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29466227](https://our.internmc.facebook.com/intern/diff/D29466227)

[ghstack-poisoned]

Varal7 committed:
Update on "Apply saved tensor hooks"


Summary: Fixes #58512

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29466227](https://our.internmc.facebook.com/intern/diff/D29466227)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29466227).

Varal7 committed:
Update on "Apply saved tensor hooks"


Summary: Fixes #58512

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29466227](https://our.internmc.facebook.com/intern/diff/D29466227)

[ghstack-poisoned]

Varal7 committed:
Update on "Apply saved tensor hooks"


Summary: Fixes #58512

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29466227](https://our.internmc.facebook.com/intern/diff/D29466227)

[ghstack-poisoned]

Varal7 committed:
Update on "Apply saved tensor hooks"


Summary: Fixes #58512

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29466227](https://our.internmc.facebook.com/intern/diff/D29466227)

[ghstack-poisoned]

Varal7 committed:
Update on "Apply saved tensor hooks"


Summary: Fixes #58512

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29466227](https://our.internmc.facebook.com/intern/diff/D29466227)

[ghstack-poisoned]

Varal7 committed:
Update on "Apply saved tensor hooks"


Summary: Fixes #58512

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29466227](https://our.internmc.facebook.com/intern/diff/D29466227)

[ghstack-poisoned]

Varal7 committed:
Update on "Apply saved tensor hooks"


Summary: Fixes #58512

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29466227](https://our.internmc.facebook.com/intern/diff/D29466227)

[ghstack-poisoned]

Varal7 committed:
Update on "Apply saved tensor hooks"


Summary: Fixes #58512

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29466227](https://our.internmc.facebook.com/intern/diff/D29466227)

[ghstack-poisoned]

Varal7 committed:
Update on "Apply saved tensor hooks"


Summary: Fixes #58512

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29466227](https://our.internmc.facebook.com/intern/diff/D29466227)

[ghstack-poisoned]

Varal7 committed:
Update on "Apply saved tensor hooks"


Summary: Fixes #58512

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29466227](https://our.internmc.facebook.com/intern/diff/D29466227)

[ghstack-poisoned]

Varal7 committed:
Update on "Apply saved tensor hooks"


Summary: Fixes #58512

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29466227](https://our.internmc.facebook.com/intern/diff/D29466227)

[ghstack-poisoned]

Varal7 committed:
Update on "Apply saved tensor hooks"


Summary: Fixes #58512

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29466227](https://our.internmc.facebook.com/intern/diff/D29466227)

[ghstack-poisoned]

Varal7 committed:
Update on "Apply saved tensor hooks"


Summary: Fixes #58512

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29466227](https://our.internmc.facebook.com/intern/diff/D29466227)

[ghstack-poisoned]

Varal7 committed:
Update on "Apply saved tensor hooks"


Summary: Fixes #58512

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29466227](https://our.internmc.facebook.com/intern/diff/D29466227)

[ghstack-poisoned]

Varal7 committed:
Update on "Apply saved tensor hooks"


Summary: Fixes #58512

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29466227](https://our.internmc.facebook.com/intern/diff/D29466227)

[ghstack-poisoned]

Varal7 committed:
Update on "Apply saved tensor hooks"


Summary: Fixes #58512

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29466227](https://our.internmc.facebook.com/intern/diff/D29466227)

[ghstack-poisoned]

Varal7 committed:
Update on "Apply saved tensor hooks"


Summary: Fixes #58512

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29466227](https://our.internmc.facebook.com/intern/diff/D29466227)

[ghstack-poisoned]

Varal7 committed:
Update on "Apply saved tensor hooks"


Summary: Fixes #58512

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29466227](https://our.internmc.facebook.com/intern/diff/D29466227)

[ghstack-poisoned]

Varal7 committed:
Update on "Apply saved tensor hooks"


Summary: Fixes #58512

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29466227](https://our.internmc.facebook.com/intern/diff/D29466227)

[ghstack-poisoned]

Varal7 committed:
Update on "Apply saved tensor hooks"


Summary: Fixes #58512

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29466227](https://our.internmc.facebook.com/intern/diff/D29466227)

[ghstack-poisoned]

albanD commented on torch/csrc/autograd/python_saved_variable_hooks.cpp:
```diff
@@ -9,19 +11,35 @@ namespace torch { namespace autograd {
-    TORCH_CHECK_NOT_IMPLEMENTED(false, "Hooks are not implemented yet");
+    py::gil_scoped_acquire acquire;
+    auto unpack_hook = py::reinterpret_borrow<py::function>(unpack_hook_);
+    py::object obj = py::cast<py::object>(data_); // copies the content of data_
```
This doesn't actually perform a copy does it? It does get a new reference though?

Varal7 replied:
Yes, sorry my bad, it's only a copy in the sense that we get a new reference.

albanD commented on torch/csrc/autograd/python_saved_variable_hooks.cpp:
```diff
@@ -9,19 +11,35 @@ namespace torch { namespace autograd {
   void PySavedVariableHooks::call_pack_hook(at::Tensor &tensor) {
-    TORCH_CHECK_NOT_IMPLEMENTED(false, "Hooks are not implemented yet");
+    py::gil_scoped_acquire acquire;
+    auto pack_hook = py::reinterpret_borrow<py::function>(pack_hook_); // borrows the reference to the hook
```
Do you plan to leave all the ref counting comments?

Varal7 replied:
I tried simplifying the ref counting comments

albanD commented on torch/csrc/autograd/saved_variable.cpp:
```diff
@@ -171,6 +209,37 @@ void SavedVariable::register_hooks(std::unique_ptr<SavedVariableHooks>&& hooks)
+    data_ = data_.tensor_data();
+  }
+
+  const bool prev = GradMode::is_enabled();
```
You can use `NoGradGuard guard;` to disable autograd as long as that variable exists.

albanD commented on torch/csrc/autograd/saved_variable.h:
```diff
@@ -55,6 +53,15 @@ class TORCH_API SavedVariable {
+  // that case was_default_constructed_ will be kept at true
+  // 2. The saved variable has been released by calling SavedVariable::reset_data(), typically
+  // during the backward pass
+  // 3. Hooks have been registered. In that case, hooks_ and py_data_ will be defined instead.
```
nit py_data_ doesn't exist anymore

albanD reviewed (commented):
The high level logic looks good. Just some cleanup to do.

Varal7 committed:
Update on "Apply saved tensor hooks"


Summary: Fixes #58512

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29466227](https://our.internmc.facebook.com/intern/diff/D29466227)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29466227).

Varal7 committed:
Update on "Apply saved tensor hooks"


Summary: Fixes #58512
Uses the hooks introduced in #60663: upon registering, the pack hook is called and the returned python object is stored. From then on, whenever we need to unpack it, we will use that python object in combination with the unpack hook.


Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29466227](https://our.internmc.facebook.com/intern/diff/D29466227)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29466227).

Varal7 committed:
Update on "Apply saved tensor hooks"


Summary: Fixes #58512
Uses the hooks introduced in #60663: upon registering, the pack hook is called and the returned python object is stored. From then on, whenever we need to unpack it, we will use that python object in combination with the unpack hook.
The packing can be done with gradient tracking disabled as we will add back the correct grad_fn during unpacking.


Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29466227](https://our.internmc.facebook.com/intern/diff/D29466227)

[ghstack-poisoned]

Varal7 committed:
Update on "Apply saved tensor hooks"


Summary: Fixes #58512
Uses the hooks introduced in #60663: upon registering, the pack hook is called and the returned python object is stored. From then on, whenever we need to unpack it, we will use that python object in combination with the unpack hook.
The packing can be done with gradient tracking disabled as we will add back the correct grad_fn during unpacking.


Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29466227](https://our.internmc.facebook.com/intern/diff/D29466227)

[ghstack-poisoned]

Varal7 committed:
Update on "Apply saved tensor hooks"


Summary: Fixes #58512
Uses the hooks introduced in #60663: upon registering, the pack hook is called and the returned python object is stored. From then on, whenever we need to unpack it, we will use that python object in combination with the unpack hook.
The packing can be done with gradient tracking disabled as we will add back the correct grad_fn during unpacking.


Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29466227](https://our.internmc.facebook.com/intern/diff/D29466227)

[ghstack-poisoned]

soulitzer commented on torch/csrc/autograd/saved_variable.cpp:
```diff
@@ -57,89 +60,126 @@ SavedVariable::SavedVariable(const Variable& variable, bool is_output, bool is_i
+  // alive, the accumulator should be kept alive by the references in the
+  // graph.
+  if (requires_grad_ && !var.grad_fn() && grad_accumulator_.expired()) {
+    TORCH_CHECK(false, "No grad accumulator for a saved leaf!");
```
Should this be TORCH_INTERNAL_ASSERT, or is it possible for user code to reach this point (in which case we could add a test for this?)

Varal7 replied:
This code is not (should not be) reachable from the user.  Will change to TORCH_INTERNAL_ASSERT.

soulitzer reviewed (commented):
Looks mostly good! Just would like to see a couple more cases tested.

soulitzer commented on torch/csrc/autograd/saved_variable.cpp:
```diff
@@ -57,89 +60,126 @@ SavedVariable::SavedVariable(const Variable& variable, bool is_output, bool is_i
+                                               : grad_fn_;
+
+  if (!is_leaf_ && !grad_fn) {
+    TORCH_CHECK(saved_for, "No grad_fn for non-leaf saved variable");
```
Same for this, is this reachable by the user?

Varal7 replied:
This code is not (should not be) reachable from the user.  Will change to TORCH_INTERNAL_CHECK.

soulitzer commented on torch/csrc/autograd/saved_variable.cpp:
```diff
@@ -154,6 +194,9 @@ Variable SavedVariable::unpack(std::shared_ptr<Node> saved_for) const {
 }
 
 void SavedVariable::register_hooks(std::unique_ptr<SavedVariableHooks>&& hooks) {
+  TORCH_CHECK(!hooks_,
```
Would be good to see a test for this also!

Varal7 replied:
This check was added with the previous PR so it was tested back then (https://github.com/pytorch/pytorch/pull/60663/files#diff-964311d1ab760238d76cb29a0e25e64f2c18cb3dbb076e10d6337c59d6dd718aR4816)

soulitzer replied:
Oh whoops, sorry

Varal7 commented on torch/csrc/autograd/saved_variable.cpp:
```diff
@@ -167,10 +210,28 @@ void SavedVariable::register_hooks(std::unique_ptr<SavedVariableHooks>&& hooks)
   hooks_ = std::move(hooks);
+
+  // If we didn't save the original variable, we already have all we need to reconstruct it
+  if (saved_original_) {
```
Sounds good, thanks!

Varal7 committed:
Update on "Apply saved tensor hooks"


Summary: Fixes #58512
Uses the hooks introduced in #60663: upon registering, the pack hook is called and the returned python object is stored. From then on, whenever we need to unpack it, we will use that python object in combination with the unpack hook.
The packing can be done with gradient tracking disabled as we will add back the correct grad_fn during unpacking.
Inplace operations done by the pack_hook on the original tensor (in the case `leaf || !output`) will be caught if the Saved tensor is used by another op.
Inplace operations done by unpack_hook will unfortunately not be caught. We will add a warning in the docs that follows this PR.


Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29466227](https://our.internmc.facebook.com/intern/diff/D29466227)

[ghstack-poisoned]

Varal7 committed:
Update on "Apply saved tensor hooks"


Summary: Fixes #58512
Uses the hooks introduced in #60663: upon registering, the pack hook is called and the returned python object is stored. From then on, whenever we need to unpack it, we will use that python object in combination with the unpack hook.
The packing can be done with gradient tracking disabled as we will add back the correct grad_fn during unpacking.
Inplace operations done by the pack_hook on the original tensor (in the case `leaf || !output`) will be caught if the Saved tensor is used by another op.
Inplace operations done by unpack_hook will unfortunately not be caught. We will add a warning in the docs that follows this PR.


Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29466227](https://our.internmc.facebook.com/intern/diff/D29466227)

[ghstack-poisoned]

Varal7 commented:
@soulitzer Thanks for the review. I added tests for three cases:
- Leaf
- Not output
- Output and not leaf (so that `saved_original` is true).
In each case, I check that either the `grad_fn` is correct or the gradient accumulated correctly, depending whether it's a leaf or not.

Varal7 committed:
Update on "Apply saved tensor hooks"


Summary: Fixes #58512
Uses the hooks introduced in #60663: upon registering, the pack hook is called and the returned python object is stored. From then on, whenever we need to unpack it, we will use that python object in combination with the unpack hook.
The packing can be done with gradient tracking disabled as we will add back the correct grad_fn during unpacking.
Inplace operations done by the pack_hook on the original tensor (in the case `leaf || !output`) will be caught if the Saved tensor is used by another op.
Inplace operations done by unpack_hook will unfortunately not be caught. We will add a warning in the docs that follows this PR.


Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29466227](https://our.internmc.facebook.com/intern/diff/D29466227)

[ghstack-poisoned]

soulitzer commented on torch/csrc/autograd/saved_variable.cpp:
```diff
@@ -57,89 +60,126 @@ SavedVariable::SavedVariable(const Variable& variable, bool is_output, bool is_i
+  // should have saved the grad accumulator. Even if the Variable is no longer
+  // alive, the accumulator should be kept alive by the references in the
+  // graph.
+  if (requires_grad_ && !var.grad_fn() && grad_accumulator_.expired()) {
```
nit: move this logic into the assert?

soulitzer reviewed (approved):
Thanks for adding the tests!

Varal7 committed:
Update on "Apply saved tensor hooks"


Summary: Fixes #58512
Uses the hooks introduced in #60663: upon registering, the pack hook is called and the returned python object is stored. From then on, whenever we need to unpack it, we will use that python object in combination with the unpack hook.
The packing can be done with gradient tracking disabled as we will add back the correct grad_fn during unpacking.
Inplace operations done by the pack_hook on the original tensor (in the case `leaf || !output`) will be caught if the Saved tensor is used by another op.
Inplace operations done by unpack_hook will unfortunately not be caught. We will add a warning in the docs that follows this PR.


Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29466227](https://our.internmc.facebook.com/intern/diff/D29466227)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29466227).
